packages/hardhat-core/src/internal/core/flamegraph.ts

```diff
@@ -0,0 +1,259 @@
+import fs from "fs";
+import { assertHardhatInvariant } from "./errors";
+import { TaskProfile } from "./task-profiling";
+
+export interface Flamegraph {
+  name: string;
+  value: number;
+  children: Flamegraph[];
+  parallel: boolean;
+}
+
+export function profileToFlamegraph(profile: TaskProfile): Flamegraph {
+  assertHardhatInvariant(
+    profile.end !== undefined,
+    `Formatting invalid task profile for ${profile.name}. No end was recorded.`
+  );
+
+  return {
+    name: profile.name,
+    // We assume this is a safe int, which is ok unless a task runs for months
+    value: Number(profile.end - profile.start).valueOf(),
+    children: profile.children.map((c) => profileToFlamegraph(c)),
+    parallel: profile.parallel === true,
+  };
+}
+
+/**
+ * Merges compatible children of toFold and its children.
+ *
+ * Compatible in a traditional Flamegraph means having the same name. We
+ * modified that notion and also require their `parallel` flag to have the same
+ * value. This means that parallel and non-parallel calls to the same function
+ * are shown with two Flamegraph "blocks".
+ *
+ * The parallel block shows the max running time, instead of the sum of them.
+ **/
+function foldFramegraph(toFold: Flamegraph): Flamegraph {
+  if (toFold.children.length === 0) {
+    return {
+      name: toFold.name,
+      value: toFold.value,
+      children: [],
+      parallel: toFold.parallel,
+    };
+  }
+
+  const children = toFold.children.map((c) => foldFramegraph(c));
+  children.sort((a, b) => (a.parallel === b.parallel ? 0 : 1));
```

Comment:
The intended behavior is similar to what you proposed. But take into account that `Array.prototype.sort` is a stable sort.

We first sort by parallel, then by name. So that all of the children with the same name are together, and within a name-group, the parallel ones are also together.

Reply:
Please ping me on slack if you want to go over this together.

Reply:
I understand this now. What I didn't understand was that the goal is to have the items grouped by name, and sorted internally by parallel, not the other way around (sorted by parallel state, and internally by name).

We need to fix the sort by parallel to return a `-1` too though.